metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractClass.scala

```diff
@@ -0,0 +1,266 @@
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Defn
+import scala.meta.Import
+import scala.meta.Mod
+import scala.meta.Pkg
+import scala.meta.Source
+import scala.meta.Stat
+import scala.meta.Tree
+import scala.meta.Type
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals._
+import scala.meta.internal.parsing.Trees
+import scala.meta.pc.CancelToken
+import scala.meta.transversers.SimpleTraverser
+
+import org.eclipse.{lsp4j => l}
+
+class ExtractClass(
+    buffers: Buffers,
+    trees: Trees
+) extends CodeAction {
+
+  override def contribute(params: l.CodeActionParams, token: CancelToken)(
+      implicit ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = {
+    val uri = params.getTextDocument.getUri
+    val path = uri.toAbsolutePath
+    val range = params.getRange
+
+    Future {
+      trees.get(path) match {
+        case Some(tree) =>
+          // Get the file name without extension
+          val fileName = uri.toAbsolutePath.filename.replaceAll("\\.scala$", "")
+
+          val definitions = classDefinitions(tree)
+          val sldNames = sealedNames(tree)
+
+          val classAtCursor =
+            definitions.find(_.name.pos.toLSP.overlapsWith(range))
+          val differentNames = classAtCursor.exists(_.name.value != fileName)
+          val notExtendsSealedOrSealed =
+            classAtCursor.exists(cls => notSealed(cls, sldNames))
+
+          val canRenameClass = differentNames &&
+            definitions.size == 1 &&
+            !getExtendsNames(definitions.head).contains(fileName)
+
+          val canExtractClass = differentNames && definitions.size > 1 &&
+            notExtendsSealedOrSealed
+
+          // This check is here only to avoid doing more computations in case of a single class in the file
+          if (canExtractClass) {
+            val classDefn = classAtCursor.get
+            val newTree =
+              getNewFileTreeStructure(tree, range, classDefn, fileName)
+            extractClassAction(uri, newTree, classDefn) :: Nil
+          } else if (canRenameClass) {
+            List(renameFileAsClassAction(uri, definitions.head))
+          } else Nil
+
+        case _ => Nil
+      }
+
+    }
+  }
+
+  private def classDefinitions(tree: Tree): List[Defn.Class] = {
+    val nodes: ListBuffer[Defn.Class] = ListBuffer()
+
+    val traverser = new SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg =>
+          super.apply(p)
+        case classNode: Defn.Class =>
+          nodes += classNode
+        case s: Source =>
+          super.apply(s)
+        case _ =>
+      }
+    }
+    traverser(tree)
+
+    nodes.toList
+  }
+
+  private def sealedNames(tree: Tree) = tree.collect {
+    case node: Defn.Trait if node.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
+      node.name.value
+    case node: Defn.Class if node.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
```

Comment:
Can enum be extended and sealed?

Reply:
Scala 2 enums can be sealed being just objects with Enumeration, but Scala 3 enums cannot. 
The latest commit should consider this.